sphinx/application.py

```diff
@@ -310,23 +311,34 @@ def _init_env(self, freshenv):
                     logger.info(_('failed: %s'), err)
                 self._init_env(freshenv=True)
 
-    def _init_builder(self, buildername):
-        # type: (unicode) -> None
+    def create_builder(self, buildername):
+        # type: (unicode) -> Builder
         if buildername is None:
             logger.info(_('No builder selected, using default: html'))
             buildername = 'html'
         if buildername not in self.builderclasses:
             raise SphinxError(_('Builder name %s not registered') % buildername)
 
         builderclass = self.builderclasses[buildername]
-        self.builder = builderclass(self)
+        return builderclass(self)
+
+    def _init_builder(self):
+        # type: () -> None
+        self.builder.set_environment(self.env)
+        self.builder.init()
         self.emit('builder-inited')
 
     def _init_enumerable_nodes(self):
         # type: () -> None
         for node, settings in iteritems(self.enumerable_nodes):
             self.env.get_domain('std').enumerable_nodes[node] = settings  # type: ignore
 
+    @property
+    def buildername(self):
+        # type: () -> unicode
+        warnings.warn('app.buildername is deprecated.', RemovedInSphinx17Warning)
```

Comment:
It would be better if the warning message describe how do extensions get builder name. i.e. `Please use app.builder.name instead.`